kad-dht/README.md

```diff
@@ -0,0 +1,427 @@
+  key (`GET_VALUE`).
+- **Content routing** - _Adding_ oneself to the list of providers for a given
+  key at the nodes closest to that key (`ADD_PROVIDER`) and _getting_ providers
+  for a given key from the nodes closest to that key (`GET_PROVIDERS`).
+
+In addition the libp2p Kademlia DHT offers the auxiliary _bootstrap_ operation.
+
+### Peer routing
+
+The below is one possible algorithm to find nodes closest to a given key on
+the DHT. Implementations may diverge from this base algorithm as long as they
+continue to adhere to the wire format.
+
+Let's assume we’re looking for nodes closest to key `Key`. We then enter an
+iterative network search.
+
+We keep track of the set of peers we've already queried (`Pq`) and the set of
+next query candidates sorted by distance from `Key` in ascending order (`Pn`).
+At initialization `Pn` is seeded with the `α` peers from our routing table we
+know are closest to `Key`, based on the XOR distance function (see [distance
+definition](#distance)).
+
+Then we loop:
+
+1. > The lookup terminates when the initiator has queried and gotten responses
+   from the k (see [#replication-parameter-k]) closest nodes it has seen.
+
+   (See Kademlia paper [0].)
+
+   The lookup might terminate early in case the local node queried all known
+   nodes, with the number of nodes being smaller than `k`.
+2. Pick as many peers from the candidate peers (`Pn`) as the `α` concurrency
+   factor allows. Send each a `FIND_NODE(Key)` request, and mark it as _queried_
+   in `Pq`.
+3. Upon a response:
+	1. If successful the response will contain the `k` closest nodes the peer
+       knows to the key `Key`. Add them to the candidate list `Pn`, except for
+       those that have already been queried.
+	2. If an error or timeout occurs, discard it.
+4. Go to 1.
+
+### Value routing
+
+Value routing can be used both to (1) _put_ and _get_ arbitrary values and (2)
+to _put_ and _get_ the public keys of nodes. Node public keys are stored in
+records under the `/pk` namespace. That is, the entry `/pk/<peerID>` will store
+the public key of peer `peerID`.
+
+DHT implementations may optimise public key lookups by providing a
+`GetPublicKey(peer.ID) (ci.PubKey)` method, that, for example, first checks if
+the key exists in the local peerstore.
+
+The lookup for public key values is identical to the lookup of arbitrary values,
+except that a custom value validation strategy is applied. It checks that
+equality `SHA256(value) == peerID` stands when:
+
+1. Receiving a response from a `GET_VALUE` lookup.
+2. Storing a public key in the DHT via `PUT_VALUE`.
+
+The record is rejected if the validation fails.
+
+#### Putting values
+
+To _put_ a value the DHT finds `k` or less closest peers to the key of the value
+using the `FIND_NODE` RPC (see [peer routing section](#peer-routing)), and then
+sends a `PUT_VALUE` RPC message with the record value to each of the peers.
+
+#### Getting values
```

Comment:
1. Most of this logic should go into a "query" section where we discuss how we find the closest peers.
2. This is now completely wrong. We (a) find the closest peers then (b) stop. We don't care about the number of values, that's not relevant.

Reply:
> where we discuss how we find the closest peers.

This is described in `### Peer routing`.

> This is now completely wrong. We (a) find the closest peers then (b) stop. We don't care about the number of values, that's not relevant.

I am not sure I follow. Could you expand? Just to avoid confusion, the procedure to find the closest peers is described in `### Peer routing`. Are you implying that go-libp2p-kad-dht dropped the notion of quorums for value lookups? rust-libp2p still offers specifying a quorum when looking up a value which is relevant outside of the ipfs space e.g. in Polkadot.